pkg/sink/postgres.go

```diff
@@ -561,29 +580,92 @@ func (p *PGXSink) handleUpdate(ctx context.Context, m *pb.Change) (err error) {
 )
 
-func (p *PGXSink) handleCommit(cp cursor.Checkpoint, commit *pb.Commit) (err error) {
-	ctx := context.Background()
-	if err = p.flushInsert(ctx); err != nil {
+func (p *PGXSink) handleCommit(sourceRemaining int, change source.Change, commit *pb.Commit) (err error) {
+	if err = p.flushInsert(); err != nil {
 		return err
 	}
-	if _, err = p.conn.Prepare(ctx, UpdateSourceSQL, UpdateSourceSQL); err != nil {
+
+	var (
+		cmt   []byte
+		seq   []byte
+		mid   []byte
+		cmtTs []byte
+		id    []byte
+	)
+
+	cp := change.Checkpoint
+	// pgtype does not support uint64, so we have to encode it as text
+	cmt, err = p.conn.TypeMap().Encode(0, pgtype.TextFormatCode, pgLSN(cp.LSN), nil)
+	if err != nil {
+		return err
+	}
+	seq, err = p.conn.TypeMap().Encode(pgtype.Int4OID, pgtype.BinaryFormatCode, pgInt4(int32(cp.Seq)), nil)
+	if err != nil {
+		return err
+	}
+	mid, err = p.conn.TypeMap().Encode(pgtype.ByteaOID, pgtype.BinaryFormatCode, cp.Data, nil)
+	if err != nil {
+		return err
+	}
+	cmtTs, err = p.conn.TypeMap().Encode(pgtype.TimestamptzOID, pgtype.BinaryFormatCode, pgTz(commit.CommitTime), nil)
+	if err != nil {
+		return err
+	}
+	id, err = p.conn.TypeMap().Encode(pgtype.TextOID, pgtype.BinaryFormatCode, p.pgSrcID, nil)
+	if err != nil {
 		return err
 	}
-	commitTs := pgTz(commit.CommitTime)
-	if _, err = p.conn.Exec(ctx, UpdateSourceSQL, pgLSN(cp.LSN), pgInt4(int32(cp.Seq)), cp.Data, commitTs, p.pgSrcID); err != nil {
+
+	for _, q := range p.queryParams {
+		p.pipeline.SendQueryParams(q.sql, q.args, q.paramOIDs, q.paramFormats, q.resultFormats)
+	}
+	p.pipeline.SendQueryParams(UpdateSourceSQL, [][]byte{cmt, seq, mid, cmtTs, id}, []uint32{0, pgtype.Int4OID, pgtype.ByteaOID, pgtype.TimestamptzOID, pgtype.TextOID}, []int16{pgtype.TextFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode}, []int16{pgtype.TextFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode})
+	p.queryParams = p.queryParams[:0]
+	p.currentCommit = commit
+	p.pendingCommitted = append(p.pendingCommitted, change.Checkpoint)
+
+	if len(p.pendingCommitted) == p.BatchTXSize || sourceRemaining == 0 {
+		err = p.endPipeline()
+	}
+	return
+}
+
+func (p *PGXSink) startPipeline() {
+	if p.pipeline == nil {
+		p.pipeline = p.raw.StartPipeline(context.Background())
+	}
+}
+
+func (p *PGXSink) endPipeline() (err error) {
+	if err = p.pipeline.Sync(); err != nil {
 		return err
 	}
-	if _, err = p.conn.Exec(ctx, "commit"); err != nil {
+	if err = p.pipeline.Close(); err != nil {
 		return err
 	}
-	atomic.StoreInt64(&p.replLag, time.Since(commitTs.Time).Milliseconds())
+	p.pipeline = nil
+
+	if p.currentCommit != nil {
+		atomic.StoreInt64(&p.replLag, time.Since(pgTz(p.currentCommit.CommitTime).Time).Milliseconds())
+	}
+	for _, cp := range p.pendingCommitted {
+		p.committed <- cp
```

Comment:
Is it possible to use Cumulative Ack of pulsar here? So that we can only track the last check point.

Reply:
Actually, the source is `pulsarReaderSource`. So there is no ack operation.
And I am not sure that it is a good choice to only send last check point for this `sourceToSink` processing.